lib/matchSourceExpression.js

```diff
@@ -0,0 +1,180 @@
+var urlModule = require('url');
+
+var ipv4DigitRegExp = /([0-9]|1?[0-9][0-9]|2[0-4][0-9]|25[0-5])/;
+var ipv4RegExp = new RegExp(
+    '^' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '$'
+);
+
+var defaultPortByScheme = {
+    http: 80,
+    https: 443
+};
+
+function origin(url) {
+    if (/file:\/\//.test(url)) {
+        return 'file:';
+    }
+    var matchOrigin = url.match(/^([a-z][a-z0-9+.-]+):\/\/(\*|(?:\*\.)?[a-z0-9-]+(?:\.[a-z0-9-]+)*)(?::(\d+|\*))?(\/.*)?$/i);
```

Comment:
I don't know if such a module exists? I typed them out by hand from the [grammar](https://www.w3.org/TR/CSP/#source-list-syntax). I could split them up into fragments that resemble the productions, then combine them one-validation style.